src/main/java/no/sysco/middleware/prometheus/kafka/common/KafkaClientJmxCollector.java

```diff
@@ -0,0 +1,169 @@
+package no.sysco.middleware.prometheus.kafka.common;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CounterMetricFamily;
+import io.prometheus.client.GaugeMetricFamily;
+import org.apache.kafka.common.MetricName;
+
+import javax.management.*;
+import java.util.*;
+
+// todo: doc
+// todo: add support for metric group `app-info`
+public abstract class KafkaClientJmxCollector {
+    private MBeanServer mBeanServer;
+    private String domainName;
+
+    public KafkaClientJmxCollector(MBeanServer mBeanServer, String domainName) {
+        this.mBeanServer = mBeanServer;
+        this.domainName = domainName;
+    }
+
+    /**
+     * standard JMX MBean name in the following format domainName:type=metricType,key1=val1,key2=val2
+     * example:
+     *  String objectNameWithDomain = "kafka.producer" + ":type=" + "producer-metrics" + ",client-id="+clientId;
+     * */
+    private ObjectName getObjectNameFromString(final String metricType, final String id) {
+        String objectNameWithDomain = domainName + ":type=" + metricType + ",client-id="+id;
+        System.out.println(objectNameWithDomain);
+        ObjectName responseObjectName = null;
+        try {
+            ObjectName mbeanObjectName = new ObjectName(objectNameWithDomain);
+            Set<ObjectName> objectNames = mBeanServer.queryNames(mbeanObjectName, null);
+            for (ObjectName object: objectNames) {
+                responseObjectName = object;
+            }
+        } catch (MalformedObjectNameException mfe) {
+            throw new IllegalArgumentException(mfe.getMessage());
+        }
+        return responseObjectName;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends Number> T getMBeanAttributeValue(final String metricType, final String attribute, final String id, final Class<T> returnType) {
+
+        System.out.println(String.format("domainName:%s ; metricType:%s ; attribute:%s ; client-id:%s", domainName, metricType, attribute, id));
+        ObjectName objectName = getObjectNameFromString(metricType, id);
+        if (objectName == null) {
+            // This also indicates that the mbeanObject is not registered.
+            // Check to see if it is an exceptions-$class object
+            if (metricType.startsWith("exceptions")) {
+                // The exceptions object is not added to the MBeanServer, as there
+                // were no HTTP client exceptions raised yet. Return 0.
+                if (returnType.equals(Double.class)) {
+                    return (T) Double.valueOf(0d);
```

Comment:
is boxing needed here?

Reply:
yes, because metrics have different types. Instead of creating several methods, all in one.